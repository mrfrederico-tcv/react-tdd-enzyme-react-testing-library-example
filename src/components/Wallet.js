import React, { Component } from 'react'
import { connect } from 'react-redux'
import styled from 'styled-components'

import { deposit, withdraw } from '../actions/balance'

const Root = styled.div``
const Title = styled.h3``
const Separator = styled.br``
const Input = styled.input``
const Button = styled.button``

export class Wallet extends Component {
  state = {
    balance: 0,
  }

  handleUpdateBalance = ({ target: { value } }) => {
    this.setState({ balance: parseInt(value, 10) })
  }

  handleDeposit = () => {
    const { balance } = this.state
    const { deposit } = this.props

    deposit(balance)
  }

  handleWithdraw = () => {
    const { balance } = this.state
    const { withdraw } = this.props

    withdraw(balance)
  }

  render() {
    const { balance } = this.props

    return (
      <Root>
        <Title data-testid="title">Wallet balance: {balance}</Title>
        <Separator />
        <Input
          aria-label="deposit-withdraw-input"
          onChange={this.handleUpdateBalance}
        />
        <Button role="deposit" onClick={this.handleDeposit}>
          Deposit
        </Button>
        <Button role="withdraw" onClick={this.handleWithdraw}>
          Withdraw
        </Button>
      </Root>
    )
  }
}

export default connect(
  ({ balance }) => ({ balance }),
  { deposit, withdraw },
)(Wallet)
