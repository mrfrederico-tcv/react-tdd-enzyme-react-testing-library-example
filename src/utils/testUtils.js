import React from 'react'
import { render } from '@testing-library/react'
import { createStore, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'

import rootReducer from '../reducers'

let __STORE__

const AllTheProviders = ({ children }) => {
  return <Provider store={__STORE__}>{children}</Provider>
}

const customRender = (
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...options
  } = {},
) => {
  __STORE__ = store

  return {
    ...render(ui, { wrapper: AllTheProviders, ...options }),
    store,
  }
}

export * from '@testing-library/react'
// override render method
export { customRender as render }
