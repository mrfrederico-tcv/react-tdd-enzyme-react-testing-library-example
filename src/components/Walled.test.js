import 'jest-styled-components'
import React from 'react'
import { Wallet } from './Wallet'
import { render, fireEvent } from '../utils/testUtils'

const mockDeposit = jest.fn()
const mockWithdrawal = jest.fn()
const setup = () => {
  const props = { balance: 10, deposit: mockDeposit, withdraw: mockWithdrawal }

  const utils = render(<Wallet {...props} />)
  const input = utils.getByLabelText('deposit-withdraw-input')
  const depositBtn = utils.getByRole('deposit')
  const withdrawBtn = utils.getByRole('withdraw')

  return {
    ...utils,
    input,
    depositBtn,
    withdrawBtn,
  }
}

describe('Wallet', () => {
  it('renders properly', () => {
    const { container } = setup()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('displays the balance from `props`', () => {
    const { getByText } = setup()
    const TEXT = 'Wallet balance: 10'

    expect(getByText(TEXT)).toBeTruthy()
  })

  describe('when the user types into the wallet input', () => {
    const USER_BALANCE = '25'
    let utils
    beforeEach(() => {
      utils = setup()
      const EVENT = { target: { value: USER_BALANCE } }

      fireEvent.change(utils.input, EVENT)
    })

    it('should display the correct value', () => {
      expect(utils.input.value).toBe(USER_BALANCE)
    })

    describe('and the user wants to make a deposit', () => {
      it('called `deposit()` and displays the correct wallet balance', () => {
        const TEXT = 'Wallet balance: 35'

        fireEvent.click(utils.depositBtn)

        expect(mockDeposit).toHaveBeenCalledWith(parseInt(USER_BALANCE, 10))
        utils.rerender(<Wallet balance={35} />)
        expect(utils.getByText(TEXT)).toBeTruthy()
      })
    })

    describe('and the user wants to make a withdrawal', () => {
      it('dispatches the `withdraw()` it receives from props with local balance', () => {
        const TEXT = 'Wallet balance: 0'

        fireEvent.click(utils.withdrawBtn)

        expect(mockWithdrawal).toHaveBeenCalledWith(parseInt(USER_BALANCE, 10))
        utils.rerender(<Wallet balance={0} />)
        expect(utils.getByText(TEXT)).toBeTruthy()
      })
    })
  })
})
