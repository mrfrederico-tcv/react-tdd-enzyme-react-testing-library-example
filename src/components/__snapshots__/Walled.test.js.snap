// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet renders properly 1`] = `
<div
  class=""
>
  <h3
    class=""
    data-testid="title"
  >
    Wallet balance: 
    10
  </h3>
  <br
    class=""
  />
  <input
    aria-label="deposit-withdraw-input"
    class=""
  />
  <button
    class=""
    role="deposit"
  >
    Deposit
  </button>
  <button
    class=""
    role="withdraw"
  >
    Withdraw
  </button>
</div>
`;
